Reduce Alternator table name length limit to 192 and fix crash when adding stream to table with very long name

nyh opened the pull request:
Before this series, it is possible to crash Scylla (due to an I/O error) by creating an Alternator table close to the maximum name length of 222, and then enabling Alternator Streams. This series fixes this bug in two ways:

1. On a pre-existing table whose name has whatever length, enabling Streams will check if the resulting name is too long, and if it is, report a clear error instead of crashing.
2. For new tables, the limit is lowered to 192. This limit is long enough, but ensures it will be possible to enable streams on such a table. It will also always be possible to add a GSI for such a table with name up to 29 characters (if the table name is shorter, the GSI name can be longer - the sum can be up to 221 characters).

No need to backport, Alternator Streams is still an experimental feature and this patch just improves the unlikely situation of extremely long table names.

nyh commented:
This PR is in draft mode waiting for the pull request https://github.com/scylladb/scylladb/pull/24597 to get merged (it was already approved by 3 people, but was unlucky with CI). The first two commits in the draft PR are those from that other pull request and don't need to be reviewed here - and will disappear when I finally rebase.

ScyllaPiotr commented:
> This series fixes this bug in two ways:

To be in line with what I learned so far about issues and PRs in Scylla, a mirroring issue should be required for this PR, and a relevant `Fixes:` part in the commit msg / cover letter.

Also, I suggest to clarify case 1. and explicitly mention that Stream creation will **fail** in the case of [...]. The specific condition is named only in case 2., while it pertains both.

nyh commented:
> > This series fixes this bug in two ways:
> 
> To be in line with what I learned so far about issues and PRs in Scylla, a mirroring issue should be required for this PR, and a relevant `Fixes:` part in the commit msg / cover letter.

Right, I'll do it.

> 
> Also, I suggest to clarify case 1. and explicitly mention that Stream creation will **fail** in the case of [...].

I wrote "report a clear error", which I thought means "fail" :-) I'll change the wording to "fail with a clear error".

> The specific condition is named only in case 2., while it pertains both.

I don't understand what you mean... Condition for what? case 1 was about **pre-existing** tables - tables created before this patch. This means tables whose length are up to 222, which I wrote one line above. But I can repeat the 222 number. The "case 2" has different conditions...

nyh commented:
> > > This series fixes this bug in two ways:
> > 
> > 
> > To be in line with what I learned so far about issues and PRs in Scylla, a mirroring issue should be required for this PR, and a relevant `Fixes:` part in the commit msg / cover letter.
> 
> Right, I'll do it.

I ran so fast to do it, that I ended up going back in time and did it in the past: https://github.com/scylladb/scylladb/issues/24598

nyh commented:
Now that https://github.com/scylladb/scylladb/pull/24597 went in, rebased to get rid of the first two commits, and marked the pull request ready to review. Also updated the cover letter.

scylladb-promoter commented:
## Docs Preview :book:
Docs Preview for this pull request is available [here](http://previews.docs.scylladb.com/scylladb/24717)

**Changed Files:**
- [alternator/compatibility](http://previews.docs.scylladb.com/scylladb/24717/alternator/compatibility)

**Note:** This preview will be available for 30 days and will be automatically deleted after that period. You can manually trigger a new build by committing changes.

nyh commented on docs/alternator/compatibility.md:
```diff
@@ -242,13 +242,13 @@ is different, or can be configured in Alternator:
       plus two, is limited to 222 characters.
-  This means that if you create a table whose name's length is close to 222
-  characters, you may not be able to create a GSI or LSI on it.
+  So for example, if you create a table whose name 192 characters, you
```
Thanks, pushed a new version fixing this.

ScyllaPiotr commented on alternator/executor.cc:
```diff
@@ -2745,10 +2781,12 @@ future<executor::request_return_type> executor::delete_item(client_state& client
     } catch(data_dictionary::no_such_column_family&) {
+        // DynamoDB returns validation error even when table does not exist
+        // and the table name is invalid.
+        validate_table_name(table_name);
```
`and` -> `or` ?
`table name` -> `table name provided in the request`
Also, is there any risk of searching for a CDC table name without making sure first that the requested name contains only legal characters?

nyh replied:
I don't think there's any risk - find_schema() more or less looks up something in a hash table, and if it's not legal characters, that silly name will simply not be found.

ScyllaPiotr reviewed (approved):
Very nice patch! I only made a couple of suggestions to improve comments, which is always a subjective thing to complain about.

nyh commented on alternator/executor.cc:
```diff
@@ -265,6 +275,7 @@ static bool valid_table_name_chars(std::string_view name) {
 // specifies that table names "names must be between 3 and 255 characters long
 // and can contain only the following characters: a-z, A-Z, 0-9, _ (underscore), - (dash), . (dot)
 // validate_table_name throws the appropriate api_error if this validation fails.
+// However, Alternator only allows max_table_name_length characters, not 255.
```
This comment is specifically about the dynamoDB "TableName", not anything else (not IndexName, not anything internal in Scylla). I'll think on how to rephrase this comment.

nyh commented on alternator/executor.cc:
```diff
@@ -276,6 +287,27 @@ static void validate_table_name(const std::string& name) {
+        // length. To provide a more helpful error message, we assume that
+        // cdc::log_name() always adds a suffix of the same length.
+        int suffix_len = cdc::log_name(table_name).length() - table_name.length();
+        throw api_error::validation(fmt::format("Streams cannot be added if the table name is longer than {} characters.",
```
Added the table's name. The comment about "it's the streams table" is incorrect (more on that above).

nyh commented:
Pushed a new version with cosmetic improvements suggested in @ScyllaPiotr 's review and no functional changes.

ScyllaPiotr commented on alternator/executor.cc:
```diff
@@ -834,6 +863,9 @@ future<executor::request_return_type> executor::delete_table(client_state& clien
             std::optional<data_dictionary::table> tbl = p.local().data_dictionary().try_find_table(keyspace_name, table_name);
             if (!tbl) {
+                // DynamoDB returns validation error even when table does not exist
+                // and the table name is invalid.
```
> What other "table name" can this be about?

You're right, it's self-evident from the call `validate_table_name(table_name);` which table name is meant.

ScyllaPiotr commented on alternator/executor.hh:
```diff
@@ -254,7 +254,7 @@ public:
         bool = false);
 
-    static void add_stream_options(const rjson::value& stream_spec, schema_builder&, service::storage_proxy& sp);
+    static bool add_stream_options(const rjson::value& stream_spec, schema_builder&, service::storage_proxy& sp);
```
Sorry for not noticing it before. Let's explain what is the meaning of the returned value.

ScyllaPiotr reviewed (approved):
Approving again, for clarity.

nyh commented:
@scylladb/scylla-maint please consider merging. There are two approvals and tests passed.
